### PR TITLE
Improve governance

### PR DESCRIPTION
**Pull request by Fraccaman**
- Fix an overflow bug when computing voting power tally
- Add validation of proposal .json when creating a new proposal
- Fix a bug when computing total amount of nay/yay
- Improving CLI UI

### CODE REVIEW

**Comment by grarco** [comment]
Also fixes #1113 by using [tempfile](https://crates.io/crates/tempfile)

File: .gitignore [hunk 1]
```diff
@@ -27,3 +27,8 @@ wasm/*.wasm
 
 # app version string file
 apps/version.rs
+
+# Governance artifacts
```
Comment (tzemanovic) [inline .gitignore:31]: let's not do this - any files created by test(s) should be inside a temp dir

Comment (grarco) [reply .gitignore:31]: Yes, it's actually what we are now doing with 147cf381fc9fef7a05e03ddfff1de031aee99f4c, I'll revert the gitignore, thanks!

**Comment by Fraccaman** [comment]
pls update wasm

**Comment by Fraccaman** [comment]
pls restart ci

**Comment by tzemanovic** [comment]
I'm thinking we could maybe speed up the e2e proposal_submission test - in the setup (`let test = setup::network(|genesis| genesis, None)?;`), we could modify the genesis to have shorter `min_proposal_period` and `min_proposal_grace_epochs`

File: apps/src/lib/client/tx.rs [hunk 2]
```diff
@@ -532,7 +531,59 @@ pub async fn submit_init_proposal(mut ctx: Context, args: args::InitProposal) {
...
+                == 0
+        );
+        eprintln!(
+            "Invalid proposal start epoch: {} must be greater than current \
+             epoch {} and a multiple of {}",
+            proposal.voting_start_epoch,
+            current_epoch,
+            goverance_parameters.min_proposal_period
+        );
+        safe_exit(1)
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:562]: let's skip the `safe_exit` call `if args.tx.force` is set (same at the other calls below)

File: apps/src/lib/client/tx.rs [hunk 3]
```diff
@@ -675,6 +727,14 @@ pub async fn submit_vote_proposal(mut ctx: Context, args: args::VoteProposal) {
 
         match proposal_start_epoch {
             Some(epoch) => {
+                if current_epoch < epoch {
+                    eprintln!(
+                        "Current epoch {} is not greater than proposal start \
+                         epoch {}",
+                        current_epoch, epoch
+                    );
+                    safe_exit(1)
```
Comment (tzemanovic) [inline apps/src/lib/client/tx.rs:736]: also `force` check here and below

File: apps/src/lib/node/ledger/shell/finalize_block.rs [hunk 4]
```diff
@@ -60,177 +51,8 @@ where
...
-                self.storage.transfer(
-                    &m1t(),
-                    funds,
-                    &gov_address,
-                    &transfer_address,
-                );
-            }
-        }
+        let _proposals_result =
+            execute_governance_proposals(self, new_epoch, &mut response)?;
```
Comment (tzemanovic) [inline apps/src/lib/node/ledger/shell/finalize_block.rs:55]: nice, thank you for moving this out!

Comment (tzemanovic) [reply apps/src/lib/node/ledger/shell/finalize_block.rs:55]: I will do the same with PoS slashing

**Review by tzemanovic — APPROVED** [review]

**Review by tzemanovic — APPROVED** [review]
LGTM, just couple minor changes. Thanks!

**Review by grarco — APPROVED** [review]

File: apps/src/lib/node/ledger/shell/governance.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            .read_storage_key::<Epoch>(&proposal_start_epoch_key)
+            .ok_or_else(|| {
+                Error::BadProposal(
+                    id,
+                    "Invalid proposal start_epoch.".to_string(),
+                )
+            })?;
+
+        let votes =
+            get_proposal_votes(&shell.storage, proposal_start_epoch, id);
```
Comment (tzemanovic) [inline apps/src/lib/node/ledger/shell/governance.rs:60]: should the epoch given to `get_proposal_votes` and `compute_tally` (here and other places where we're using these) be end epoch? (https://specs.anoma.net/master/architecture/governance/signaling.html#tally says it should be `end_epoch`)

**Comment by tzemanovic** [comment]
In `apps/src/lib/client/rpc.rs`, the offline tally `get_proposal_offline_votes` is missing slashes for delegator's bonds. I think we need something like this: https://github.com/anoma/anoma/commit/dad75398b30539e2da8e15db970bdcc2e32ee094. I also refactored the last bit to update the maps with https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry

Ideally, this should really be hidden behind PoS API as the data from PoS is hard to query atm. I think we can solve it in future with some common interface for storage direct access and client RPC queries

**Pull request closed without merging**